src/ipa/rpi/controller/rpi/clock_recovery.cpp

```diff
@@ -0,0 +1,140 @@
+#include "clock_recovery.h"
+
+#include <cctype>
+#include <chrono>
+#include <fcntl.h>
+#include <list>
+#include <map>
+#include <strings.h>
+#include <unistd.h>
+#include <vector>
+
+#include <libcamera/base/log.h>
+
+#include "sync_status.h"
+
+using namespace std::chrono_literals;
+using namespace RPiController;
+using namespace libcamera;
+
+LOG_DEFINE_CATEGORY(RPiClockRec)
+
+
+float ClockRecovery::slope()
+{
+	//returns slope for least squares line through points
+	double sl = (error_values.size() * xysum_ - xsum_ * ysum_) / (error_values.size() * x2sum_ - xsum_ * xsum_);
+	return sl;
+}
+
+void ClockRecovery::update_values(std::chrono::microseconds correction)
+{
+	std::list<int64_t>::iterator iterror = error_values.begin();
+	std::list<int64_t> update;
+
+	for (; iterror != error_values.end(); iterror++) {
+		update.push_back((*iterror) - correction.count());
+ 	}
+	ysum_ -= (double)correction.count() * update.size();
+	xysum_ -= correction.count() * xsum_;
+	error_values = update;
+}
+
+std::chrono::microseconds ClockRecovery::trending_error(std::chrono::microseconds LastWallClock,
```

Comment:
I'm afraid libcamera will want camelCase, not CamelCase! So `lastWallClock` in this case.